### PR TITLE
[SPARK-25267][SQL][TEST] Disable ConvertToLocalRelation in the test cases of sql/core and sql/hive

### PR DESCRIPTION
**Pull request by dilipbiswal**
## What changes were proposed in this pull request?
In SharedSparkSession and TestHive, we need to disable the rule ConvertToLocalRelation for better test case coverage. 
## How was this patch tested?
Identify the failures after excluding "ConvertToLocalRelation" rule.

### CODE REVIEW

**Comment by dilipbiswal** [comment]
retest this please

File: mllib/src/test/scala/org/apache/spark/ml/recommendation/ALSSuite.scala [hunk 1]
```diff
@@ -24,12 +24,10 @@ import scala.collection.JavaConverters._
 import scala.collection.mutable
 import scala.collection.mutable.{ArrayBuffer, WrappedArray}
 import scala.language.existentials
-
 import com.github.fommil.netlib.BLAS.{getInstance => blas}
 import org.apache.commons.io.FileUtils
 import org.apache.commons.io.filefilter.TrueFileFilter
 import org.scalatest.BeforeAndAfterEach
-
```
Comment (viirya) [inline mllib/src/test/scala/org/apache/spark/ml/recommendation/ALSSuite.scala]: Revert the blank lines?

File: mllib/src/test/scala/org/apache/spark/ml/recommendation/ALSSuite.scala [hunk 2]
```diff
@@ -652,65 +653,66 @@ class ALSSuite extends MLTest with DefaultReadWriteTest with Logging {
...
-        als.fit(df.select(df("item_big").as("item"), df("user"), df("rating")))
-      }.getCause.getMessage.contains(msg))
-      assert(intercept[SparkException] {
-        als.fit(df.select(df("item_small").as("item"), df("user"), df("rating")))
-      }.getCause.getMessage.contains(msg))
-    }
-    withClue("transform should fail when ids exceed integer range. ") {
-      val model = als.fit(df)
-      def testTransformIdExceedsIntRange[A : Encoder](dataFrame: DataFrame): Unit = {
+    withSQLConf(SQLConf.OPTIMIZER_EXCLUDED_RULES.key -> "") {
```
Comment (viirya) [inline mllib/src/test/scala/org/apache/spark/ml/recommendation/ALSSuite.scala]: If we only want to disable `ConvertToLocalRelation` in sql/core and sql/hive, maybe we can set this sql conf at `MLTest`?

I don't see any usage of `withSQLConf` in ml tests. It looks a bit weird to see it here.

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/test/TestHive.scala [hunk 3]
```diff
@@ -59,7 +60,8 @@ object TestHive
         .set("spark.sql.warehouse.dir", TestHiveContext.makeWarehouseDir().toURI.getPath)
         // SPARK-8910
         .set("spark.ui.enabled", "false")
-        .set("spark.unsafe.exceptionOnMemoryLeak", "true")))
+        .set("spark.unsafe.exceptionOnMemoryLeak", "true")
+        .set(SQLConf.OPTIMIZER_EXCLUDED_RULES.key, ConvertToLocalRelation.ruleName)))
```
Comment (viirya) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/test/TestHive.scala]: Can we add a comment for this?

File: mllib/src/test/scala/org/apache/spark/ml/recommendation/ALSSuite.scala (hunk 2 above)
Comment (viirya) [reply mllib/src/test/scala/org/apache/spark/ml/recommendation/ALSSuite.scala]: Sure. :)

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/collectionOperations.scala [hunk 4]
```diff
@@ -1464,12 +1465,14 @@ case class ArrayContains(left: Expression, right: Expression)
     nullSafeCodeGen(ctx, ev, (arr, value) => {
       val i = ctx.freshName("i")
       val getValue = CodeGenerator.getValue(arr, right.dataType, i)
+      val setIsNullCode = if (nullable) s"${ev.isNull} = true;" else ""
+      val unsetIsNullCode = if (nullable) s"${ev.isNull} = false;" else ""
       s"""
       for (int $i = 0; $i < $arr.numElements(); $i ++) {
         if ($arr.isNullAt($i)) {
-          ${ev.isNull} = true;
+          ${setIsNullCode}
```
Comment (maropu) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/collectionOperations.scala]: nit: `${setIsNullCode}` -> `$setIsNullCode`
btw, you found some bugs when excluding the `ConvertToLocalRelation` rule in tests, right?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/collectionOperations.scala [hunk 5]
```diff
@@ -1464,12 +1465,14 @@ case class ArrayContains(left: Expression, right: Expression)
...
         } else if (${ctx.genEqual(right.dataType, value, getValue)}) {
-          ${ev.isNull} = false;
+          ${unsetIsNullCode}
```
Comment (maropu) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/collectionOperations.scala]: ditto

File: sql/core/src/test/scala/org/apache/spark/sql/DataFrameSuite.scala [hunk 6]
```diff
@@ -1730,9 +1730,8 @@ class DataFrameSuite extends QueryTest with SharedSQLContext {
 
   test("SPARK-9083: sort with non-deterministic expressions") {
     import org.apache.spark.util.random.XORShiftRandom
```
Comment (maropu) [inline sql/core/src/test/scala/org/apache/spark/sql/DataFrameSuite.scala]: Can we move this import into the top? (this is not related to this pr though)

File: sql/core/src/test/scala/org/apache/spark/sql/DataFrameFunctionsSuite.scala [hunk 7]
```diff
@@ -85,12 +85,12 @@ class DataFrameFunctionsSuite extends QueryTest with SharedSQLContext {
     }
 
     val df5 = Seq((Seq("a", null), Seq(1, 2))).toDF("k", "v")
-    intercept[RuntimeException] {
+    intercept[Exception] {
       df5.select(map_from_arrays($"k", $"v")).collect
```
Comment (maropu) [inline sql/core/src/test/scala/org/apache/spark/sql/DataFrameFunctionsSuite.scala]: What's the concrete exception this query throws?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/collectionOperations.scala (hunk 4 above)
Comment (gatorsmile) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/collectionOperations.scala]: Please open separate JIRAs and PRs for the codegen fixes.

File: sql/core/src/test/scala/org/apache/spark/sql/DataFrameFunctionsSuite.scala [hunk 8]
```diff
@@ -85,12 +85,12 @@ class DataFrameFunctionsSuite extends QueryTest with SharedSQLContext {
     }
 
     val df5 = Seq((Seq("a", null), Seq(1, 2))).toDF("k", "v")
-    intercept[RuntimeException] {
+    intercept[Exception] {
```
Comment (gatorsmile) [inline sql/core/src/test/scala/org/apache/spark/sql/DataFrameFunctionsSuite.scala]: Let us capture the exception and compare the error messages.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/collectionOperations.scala (hunk 4 above)
Comment (dilipbiswal) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/collectionOperations.scala]: @gatorsmile OK.. Sean.

**Comment by cloud-fan** [comment]
Thanks for looking into this!

I think we have to clean up our test framework later (after 2.4). We should identify the test cases that are actually testing the expressions, and run it with/without enabling the local relation optimization, to test both codegen and interpreted code paths.

Since the current test suites are a little messy, this will be a lot of work, to reorganize them. I'm looking forward to seeing us accomplish it in Spark 3.0!

**Comment by dilipbiswal** [comment]
@cloud-fan Thank you Wenchen. Do we want to fix the two codegen compile errors in 2.4 ? One is in ArrayContains and the other is in ArraySort.

File: sql/core/src/test/scala/org/apache/spark/sql/DataFrameSuite.scala (hunk 6 above)
Comment (dilipbiswal) [reply sql/core/src/test/scala/org/apache/spark/sql/DataFrameSuite.scala]: @maropu will do.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/collectionOperations.scala (hunk 5 above)
Comment (dilipbiswal) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/collectionOperations.scala]: @maropu will change

File: sql/core/src/test/scala/org/apache/spark/sql/DataFrameFunctionsSuite.scala (hunk 7 above)
Comment (dilipbiswal) [reply sql/core/src/test/scala/org/apache/spark/sql/DataFrameFunctionsSuite.scala]: @maropu I will double check and get back. But i think it was SparkException.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/collectionOperations.scala [hunk 9]
```diff
@@ -1464,12 +1465,14 @@ case class ArrayContains(left: Expression, right: Expression)
     nullSafeCodeGen(ctx, ev, (arr, value) => {
       val i = ctx.freshName("i")
       val getValue = CodeGenerator.getValue(arr, right.dataType, i)
+      val setIsNullCode = if (nullable) s"${ev.isNull} = true;" else ""
+      val unsetIsNullCode = if (nullable) s"${ev.isNull} = false;" else ""
       s"""
       for (int $i = 0; $i < $arr.numElements(); $i ++) {
         if ($arr.isNullAt($i)) {
```
[the remark this answers is not in the document]
Comment (dilipbiswal) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/collectionOperations.scala]: @maropu You are right !! I will try to optimize this in the other pr i am going to open. please check if you like it.

**Comment by ueshin** [comment]
@dilipbiswal Let's fix the two functions in 2.4. Could you open a pr to fix the two functions? Thanks!

**Comment by dilipbiswal** [comment]
@ueshin Thank you.. will do.

**Comment by dilipbiswal** [comment]
@ueshin Opened (https://github.com/apache/spark/pull/22314) and (https://github.com/apache/spark/pull/22315). Thank you.

File: sql/core/src/test/scala/org/apache/spark/sql/DataFrameFunctionsSuite.scala (hunk 8 above)
Comment (HyukjinKwon) [inline sql/core/src/test/scala/org/apache/spark/sql/DataFrameFunctionsSuite.scala]: Shall we also catch specific exception per https://github.com/databricks/scala-style-guide#testing-intercepting

**Comment by gatorsmile** [comment]
@dilipbiswal Any update on this PR?

**Comment by dilipbiswal** [comment]
@gatorsmile Yeah.. i will push the changes tonight for you to take a look.

**Comment by gatorsmile** [comment]
Thank you!

**Comment by dilipbiswal** [comment]
retest this please

**Comment by dilipbiswal** [comment]
cc @gatorsmile

**Comment by dilipbiswal** [comment]
Thanks a lot @gatorsmile

File: sql/core/src/test/scala/org/apache/spark/sql/DataFrameSuite.scala [hunk 10]
```diff
@@ -1729,10 +1730,8 @@ class DataFrameSuite extends QueryTest with SharedSQLContext {
   }
 
   test("SPARK-9083: sort with non-deterministic expressions") {
-    import org.apache.spark.util.random.XORShiftRandom
-
     val seed = 33
-    val df = (1 to 100).map(Tuple1.apply).toDF("i")
+    val df = (1 to 100).map(Tuple1.apply).toDF("i").repartition(1)
```
Comment (cloud-fan) [inline sql/core/src/test/scala/org/apache/spark/sql/DataFrameSuite.scala]: Sorry I didn't follow this thread closely. Why do we need these repartition(1) changes?

Comment (dilipbiswal) [reply sql/core/src/test/scala/org/apache/spark/sql/DataFrameSuite.scala]: @cloud-fan I was just trying get this test case to pass when `ConvertToLocalRelation` is enabled as well as disabled. So when this rule is active, i saw that all the calls to random.nextXXX happens in one thread. When this rule is disabled, the random values get evaluated under the `project` operator and gets called from multiple threads. Thats why i am repartitioning the data frame to enforce single threaded execution. Is this not the right thing to do ? Please let me know ..

File: sql/core/src/test/scala/org/apache/spark/sql/DataFrameFunctionsSuite.scala [hunk 11]
```diff
@@ -85,14 +85,16 @@ class DataFrameFunctionsSuite extends QueryTest with SharedSQLContext {
     }
 
     val df5 = Seq((Seq("a", null), Seq(1, 2))).toDF("k", "v")
-    intercept[RuntimeException] {
+    val msg1 = intercept[Exception] {
```
[the remark this answers is not in the document]
Comment (dilipbiswal) [reply sql/core/src/test/scala/org/apache/spark/sql/DataFrameFunctionsSuite.scala]: @HyukjinKwon Yeah... we could have caught SparkException here. My intention was to have this test case pass both when location relation optimization is on and off. Thats why i changed it a a generic exception along with verifying the error text.

File: sql/core/src/test/scala/org/apache/spark/sql/DataFrameSuite.scala (hunk 10 above)
Comment (dilipbiswal) [reply sql/core/src/test/scala/org/apache/spark/sql/DataFrameSuite.scala]: @HyukjinKwon We are leaving this optimization on for MLTest as of now. Should we open it up for TestHive and  keep it disabled it for SharedSparkSession ? cc @gatorsmile
